CleverTapSDK/EventDatabase/CTEventDatabase.m

```diff
@@ -0,0 +1,564 @@
+//
+//  CTEventDatabase.m
+//  CleverTapSDK
+//
+//  Created by Nishant Kumar on 25/10/24.
+//  Copyright © 2024 CleverTap. All rights reserved.
+//
+
+#import "CTEventDatabase.h"
+#import "CTConstants.h"
+
+@interface CTEventDatabase()
+
+@property (nonatomic, strong) CleverTapInstanceConfig *config;
+
+@end
+
+@implementation CTEventDatabase {
+    sqlite3 *_eventDatabase;
+    dispatch_queue_t _databaseQueue;
+}
+
++ (instancetype)sharedInstanceWithConfig:(CleverTapInstanceConfig *)config {
+    static CTEventDatabase *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[self alloc] initWithConfig:config];
+    });
+    return sharedInstance;
+}
+
+- (instancetype)initWithConfig:(CleverTapInstanceConfig *)config {
+    if (self = [super init]) {
+        _config = config;
+        _databaseQueue = dispatch_queue_create([[NSString stringWithFormat:@"com.clevertap.eventDatabaseQueue:%@", _config.accountId] UTF8String], DISPATCH_QUEUE_CONCURRENT);
```

Comment:
Using a concurrent queue for database access is generally discouraged unless reads and writes are separated or synchronized.
Also, all calls to this queue are dispatch_sync which negates its concurrency to be DISPATCH_QUEUE_CONCURRENT.